plugins/gnupgplugin/gpgmebase.cpp

```diff
@@ -0,0 +1,761 @@
+    gpgme_import_status_t status = result->imports;
+    const char *fpr = (const char*)0;
+    while (status) {
+        if (gpgme_err_code(status->result) != GPG_ERR_NO_ERROR) {
+            return Result(Result::CryptoPluginKeyImportError,
+                          QStringLiteral("failing importing data: %1.").arg(gpgme_strerror(status->result)));
+        }
+        if (!fpr)
+            fpr = status->fpr;
+        status = status->next;
+    }
+    if (!fpr) {
+        return Result(Result::CryptoPluginKeyImportError,
+                      QStringLiteral("no key in the imported data."));
+    }
+
+    GPGmeKey gFetchedKey(ctx, fpr);
+    if (!gFetchedKey) {
+        return Result(Result::CryptoPluginKeyImportError,
+                      QStringLiteral("Cannot import key %1 from data: %2.").arg(fpr).arg(gFetchedKey.error()));
+    }
+    gFetchedKey.toKey(importedKey, name());
+
+    return Result();
+}
+
+Result Daemon::Plugins::GnuPGPlugin::importKey(const QByteArray &data,
+                                               const QByteArray &passphrase,
+                                               const QVariantMap &customParameters,
+                                               Key *importedKey)
+{
+    QTemporaryDir tmp;
+    if (!tmp.isValid()) {
+        return Result(Result::CryptoPluginKeyImportError,
+                      QStringLiteral("cannot create temporary directory: %1.").arg(tmp.errorString()));
+    }
+    Result result = importKey(data, passphrase, customParameters, importedKey, tmp.path());
+    if (result.code() == Result::Succeeded)
+        tmp.setAutoRemove(false);
+    return result;
+}
+
+Result Daemon::Plugins::GnuPGPlugin::importAndStoreKey(const QByteArray &data,
+                                                       const Key &keyTemplate,
+                                                       const QByteArray &passphrase,
+                                                       const QVariantMap &customParameters,
+                                                       Key *keyMetadata)
+{
+    Q_UNUSED(keyTemplate);
+
+    return importKey(data, passphrase, customParameters, keyMetadata, QString());
+}
+
+Result Daemon::Plugins::GnuPGPlugin::storedKey(const Key::Identifier &identifier,
+                                               Key::Components keyComponents,
+                                               const QVariantMap &customParameters,
+                                               Key *key)
+{
+    Q_UNUSED(customParameters);
+
+    GPGmeContext ctx(m_protocol);
+    if (!ctx)
+        return Result(Result::StorageError, ctx.error());
+
+    GPGmeKey gkey(ctx, identifier.name(),
+                  (keyComponents & Key::SecretKeyData)
+                  ? GPGmeKey::Secret : GPGmeKey::Public);
+    if (!gkey) {
+        // Todo: need to set the storage error code to
+        // Sailfish::Secrets::Result::InvalidSecretError
```

Comment:
there should be a constructor for this now :-)

Reply:
I forgot to upgrade this ;)